Add networking for co-op

spykyvenator opened the pull request:
It took way too long, but I found some time to finish this.
this should build for glibc, that is all I tested at the moment.
to disable the networking just remove the -DNET flag from CFLAGS in the configure.ac file.
Please try to build and run yourself before merging ;)

example usage:
    bumprace -S -n -w -s & 
    select any 2 player mode
    bumprace -c 127.0.0.1 -n -w -s 

And you should have two screens connected over a socket :)
you could have to update the automatic generated files, these accidentally got updated with the commit -am, maybe these could be removed in a future pull request.

spykyvenator committed:
code + gitignore

spykyvenator committed:
makefile.am edit

spykyvenator committed:
makefile edit

spykyvenator committed:
prettifications, and correct offline co-op without trying to connect to server

spykyvenator committed:
set makefiles and update configure script with aclocal

spykyvenator committed:
update: makefile

spykyvenator committed:
set NET flag in configure file

spykyvenator committed:
but last commit did not need to be in the configure file but in the configure.ac file

spykyvenator committed:
add: net to README

spykyvenator committed:
remove: listening from README

spykyvenator committed:
add: guards that exclude these

spykyvenator committed:
add comment for clarity on how not to build networking into the game

spykyvenator committed:
remove some debugging

spykyvenator commented:
and just successfully did a session with glibc server and musl client :)

karlb commented on src/bumprace.c:
```diff
@@ -51,6 +57,14 @@
+
+#ifdef NET
+       int client=0,Skt=0,Port=7664;
+       //struct protoent *protoent;
```
Looks like we don't need this anymore.

karlb commented on src/bumprace.c:
```diff
@@ -922,6 +941,11 @@ void ReadCommandLine(char *argv[])
       {i++;levelset=atoi(argv[i]);printf("Levelset is set to %d\n",levelset);} else
+#ifdef NET
+    if ( (!strcmp(argv[i],"--server")) || (!strcmp(argv[i],"-S")) ){client=1;} else
+    if ( ((!strcmp(argv[i],"--client"))||(!strcmp(argv[i],"-c")) ) && (argv[++i]) ){Ip = argv[i];client=2;} else
```
Wrong command line usage should not cause crashes. E.g.:
```
src/bumprace -w -c
fish: Job 1, 'src/bumprace -w -c' terminated by signal SIGSEGV (Address boundary error)
```

karlb commented on src/bumprace.c:
```diff
@@ -922,6 +941,11 @@ void ReadCommandLine(char *argv[])
+#ifdef NET
+    if ( (!strcmp(argv[i],"--server")) || (!strcmp(argv[i],"-S")) ){client=1;} else
+    if ( ((!strcmp(argv[i],"--client"))||(!strcmp(argv[i],"-c")) ) && (argv[++i]) ){Ip = argv[i];client=2;} else
+    if ( ((!strcmp(argv[i],"--port")) || (!strcmp(argv[i],"-p"))) && (argv[++i]) ){Port = atoi(argv[i]);} else
```
The parsing of the short arg does not work as expected in the following case
```
src/bumprace -c localhost -p 7664 -w
Unknown parameter (-h for help): "7664" 
```

spykyvenator replied:
yes there already was a -p argument :)

karlb commented on src/bumprace.c:
```diff
@@ -51,6 +57,14 @@
        int final=1,fullscreen=1,pageflip=0,sound=1,precision=10,fps=0,particle=1,mode,dofadeout=1,levelset=0;
+
+#ifdef NET
+       int client=0,Skt=0,Port=7664;
```
Any reason for having the variables capitalized even though none of the others are? I know I haven't been overly consistent with the naming, but we don't have to make it worse.

karlb commented on src/bumprace.c:
```diff
@@ -51,6 +57,14 @@
     //general setting
        int final=1,fullscreen=1,pageflip=0,sound=1,precision=10,fps=0,particle=1,mode,dofadeout=1,levelset=0;
+
+#ifdef NET
```
Any reason against always compiling with network support? That's plenty of `#ifdef`s.

spykyvenator replied:
I thought it would be nice to have the option to choose to compile this into the game and it made it clear where I have worked. I can remove them if you like
I also wanted to see how to do this in the makefiles. but it is fairly trivial as I see now

karlb reviewed (commented):
Nice to see that you continued working on this!

I currently get stuck in the racer selection on both screens when trying to run a networked game. Do you have an idea what the problem might be?

spykyvenator committed:
remove capitalisation from net variables

spykyvenator committed:
arguments:
	added extra checks to arguments so that no NULL pointer is passed to strcmp
	wrong ip format should be detected and raise an error
	--port in short is now -P since -p is already taken
gameplay:
	send racernum on levelinit so that the server can choose both racers
	client can't set racer himself
accessibility:)
	when server sets single player mode, game doesn't wait for client but starts single player game

spykyvenator commented:
last commit addresses all of the above:
arguments:
	added extra checks to arguments so that no NULL pointer is passed to strcmp
	wrong ip format should be detected and raise an error
	--port in short is now -P since -p is already taken 
gameplay:
	send racernum on levelinit so that the server can choose both racers
	client can't set racer himself
accessibility:)
	when server sets single player mode, game doesn't wait for client but starts single player game

spykyvenator commented:
I tried compiling for windows, however I get sys/socket.h: No such file or directory.
https://stackoverflow.com/questions/2952733/using-sys-socket-h-functions-on-windows
I think windows support is going to be for another time

/usr/lib/mingw64-toolchain/bin/x86_64-w64-mingw32-gcc ./*.c -g -Og -DSOUND -DNET -I/usr/include/SDL -D_GNU_SOURCE=1 -D_REENTRANT -DDATADIR=\"${datarootdir}/bumprace\" -lSDL_image -ljpeg -lm  -L/usr/lib64 -lSDL
-lSDL_image -lSDL_mixer -lz -lSDL -o ./bumpracewin.exe
just so you and mostly I can see what command I used to try to compile to windows.
Another good reason to use the NET flag! :)

spykyvenator commented:
with you getting stuck on the selection menu, does the last commit change that? otherwise can I get the commands you ran?

spykyvenator committed:
add more NET ifdefs for succesful compilation without -DNET

spykyvenator committed:
apperently inline funcions dont work that well with ifdef

spykyvenator commented:
Ok I can now compile with CC=/usr/lib/mingw64-toolchain/bin/x86_64-w64-mingw32-gcc and I have a windows build without NET
that works with wine

karlb commented on src/bumprace.c:
```diff
@@ -1218,7 +1218,10 @@ if (client && (!(skt = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) )) {
   Font=LoadImage("font.png",3);
   InitFont(Font);
-  if (final && client != 2) StartText(i);
+#ifdef NET
+  if (client != 2)
+#endif //NET
```
You could set `client = 0` once if compiled without NET, then you would not need the ifdef here (and below) and could keep the single if.

spykyvenator replied:
I'Il just leave it like this, it is ugly, but it does keep all the networking code together.

spykyvenator replied:
getting the server hostaddresss is quite difficult since it can have multiple interfaces and ip addresses.

karlb reviewed (commented):
If the client connects before the server has chosen the racer, the server does not notice the client. If the client connects after the racers have been chosen, it works fine.

Apart from that, I only have some suggestions you can implement or ignore as you like (in this PR or later):
* Allow passing host name or IP address
* Show the server name or IP address when waiting for clients
* When no input is accepted show a message why (waiting for client to connect)
* Let each player choose his own racer

spykyvenator committed:
let client try to reconnect multiple times

spykyvenator committed:
fix: timeout with setsockopt

spykyvenator committed:
edit: set timeout only once

spykyvenator committed:
add: wait menu for server

spykyvenator committed:
add: menus for both client and server waiting

spykyvenator committed:
add: make server and client each select their own racer

spykyvenator committed:
prettify: NET ifdefs

spykyvenator commented:
- [Not Done] Allow passing host name or IP address => since this is making things more platform specific I first would like to get the networking working in windows.
- [Not Done] Show the server name or IP address when waiting for clients => While this also is platform specific this is even harder because of multiple interfaces and ip addresses, (which one to display)?
- [Done] When no input is accepted show a message why (waiting for client to connect)
- [Done] Let each player choose his own racer

I would like to merge now if that fits for you :)

karlb reviewed (approved):
Nice, works fine now!

I sometimes have problems quitting cleanly, but I'm not sure if that's related to the changes in this PR. I didn't run this game frequently in the last years.
